src/crimson/net/SocketConnection.cc

```diff
@@ -892,6 +884,8 @@ SocketConnection::execute_open()
 {
   logger().debug("{} trigger open, was {}", *this, static_cast<int>(state));
   state = state_t::open;
+  // satisfy the handshake's promise
+  seastar::now().forward_to(std::move(h.promise));
```

Comment:
i'd suggest fulfill the promise in a single place -- move this line back before `return dispatcher.ms_handle_accept(this);`.

and use `h.promise.set_value()` instead, which is more efficient and more straightforward -- save a temporary  `future<>` object.

Reply:
> move this line back before return dispatcher.ms_handle_accept(this);

Several considerations here:
- We need to fulfill `h.promise` for 2 different purposes: unblock and dispatch `write_message()` `do_keepalive()` in open state; unblock and drop pending `send()` `keepalive()` in the closing state.
- The fulfilling of `h.promise` should be atomic with state changes, thus `send()` `keepalive()` can correctly decide its behavior according to the current state. Whether do normal sending or just drop.

> use h.promise.set_value() instead

done.